charts/user-permissions/templates/roles-and-bindings.yaml

```diff
@@ -0,0 +1,98 @@
+{{ $roleDefinitions := .Values.roleDefinitions }}
+{{ $byNamespaceAndName := dict }}
+
+{{/*
+
+For each entry in .Values.bindings, if it's the first time we have
+seen this namespace and role name, create a (Cluster)Role for it,
+using the rules from the same-named entry in $roleDefinitions.
+
+Keep track of which subjects to bind to it in $byNamespaceAndName, for later.
+
+*/}}
+
+{{- range .Values.bindings }}
+  {{ $currentBinding := . }}
+  {{- range .bindTo }}
+    {{ $namespace := .namespace }}
+    {{ $isCluster := eq $namespace "*" }}
+    {{ $clusterStr := $isCluster | ternary "Cluster" "" }}
+    {{- range .roles }}
+      {{ $key := printf "%s/%s" $namespace . }}
+      {{ $bindings := get $byNamespaceAndName $key }}
+      {{- if not $bindings }}
+        {{ $bindings = list }}
+
+        {{/* BEGIN ROLE */}}
```

Comment:
Would it make sense / be possible to store a tuple of `{roleDefinitions, bindings}` in the map instead of just the bindings, so that the role output could be moved outside the loop that handles the logic?